### PR TITLE
Speak typed characters in Windows consoles in the correct language (second try)

### LINKED ISSUE #10113: UIA console: when typing in the console, characters are wrongly pronounced regardless of the keyboard layout
### Steps to reproduce:
1. open a console window with the uia console option turned on, for example a wsl console
. switch to croatian or to the US keyboard language. I am not aware of other layouts having the problem which i will describe here thoroughly
russian keyboard is also included there
2. try to type  or ], ; or ', \ or /
### Actual behavior:
note: here are given the corresponding keys on the us keyboard. these keys are wrongly interpreted by NVDA also on oother non-us keyboard layout
for example / slash on US keyboard is interpreted as apostrophe,
; as č, the croatian diacritic mark, and so on.
### Expected behavior:
the signs should be interpreded bz NVDA as these are written on the kezboard respectivelz
#### NVDA installed/portable/running from source:
NVDA installed
#### NVDA version:
Latest alpha
#### Windows version:
18362 latest indows 10 1903
#### Name and version of other software in use when reproducing the issue:
n-a
#### Other information about your system:
to reproduce this bug, you should have turned on have character echo turned on
#### Does the issue still occur after restarting your PC?
yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
n/a

### PR DESCRIPTION
**Pull request by michaelDCurran**
Replaces pr #10551 
 
### Link to issue number:
Fixes #10113 

### Summary of the issue:
With the work to support Windows consoles via UIA, NVDA moved to speaking typed characters in consoles by detecting keyboard input and then using ToUnicodeEx to convert the keyboard input into a character for speaking. We had already been taking this approach for Microsoft Edge and other places in Windows 10 where we could not inject in-process.
ToUnicodeEx requires a keyboard layout to use the correct language. Therefore, we would call GetKeyboardLayout, giving it the thread ID of the focused window.
However, it seems that Windows console windows report the wrong thread ID, and therefore, GetKeyboardLayout fails, returning 0 instead of a valid keyboard layout.
Passing this 0 to ToUnicodeEx, it would generate characters in the user's system language (I think). At least, on my system it was always english US no matter what input language I was currently using.
 
### Description of how this pull request fixes the issue:
Override the windowThreadID property on both UIA and legacy WinConsole NVDAObjects, to reflect the true thread ID of the window. This thread ID is found in a new cache of console windows to thread IDs in IAccessibleHandler, filled in by received winEvents for ConsoleWindowClass windows.
Received winEvents always contain an ID of the true thread that NotifyWinEvent was called in. In the case of ConsoleWindowClass winEvents, this is from inside conhost.
this thread happens to be the true input (focus thread) for consoles, and therefore GetKeyboardLayout returns the correct input language for the console, when given this threadID.
We can garantee that there will at least either be a focus winEvent, or a locationChange winEvent (for the caret) for consoles, when moving to a console and or starting NVDA within a console. This is enough to make sure that we have a threadID for that console window when we need it.

### Testing performed:
* Installed the Arabic keyboard layout.
* Opened a Windows console and ensured NVDA was using UIA.
* With speak typed characters on, first typed some characters, confirming they were being announced in english. backspacing them also confirming they were english characters written to the console.
* Switched keyboard layouts to Arabic.
* Again typed some characters, confirming that NVDA was speaking them as Arabic. Backspacing them, confirmed that actual arabic characters were written to the console.
  
### Known issues with pull request:
It is possible that the very first typed character when starting NVDA from within a console will be announced in the wrong language, as the caret has not yet moved once. But this is rather rare.

### Change log entry:
None

### CODE REVIEW

**Review by codeofdusk — APPROVED** [review]
Looks fine to me, assuming that nothing in the legacy console support depends on the previous threadID behavior.

File: source/IAccessibleHandler.py [hunk 1]
```diff
@@ -1102,3 +1106,13 @@ def isMarshalledIAccessible(IAccessibleObject):
...
 	windll.kernel32.GetModuleFileNameW(handle,buf,1024)
 	return not buf.value.lower().endswith('oleacc.dll')
+
+
+#: Maps from console windows (ConsoleWindowClass) to thread IDs
+# Windows hacks GetWindowThreadProcessId to return the input thread of the first attached process in a console
+# But NVDA really requires to know the actual thread the window was created in,
+# I.e. inside conhost,
+# In order to handle speaking of typed characters etc.
+# winEventCallback adds these when ever it sees an event for ConsoleWindowClass windows,
```
Comment (leonardder) [inline source/IAccessibleHandler.py]: Should this be whenever?

```suggestion
# winEventCallback adds these whenever it sees an event for ConsoleWindowClass windows,
```

File: source/IAccessibleHandler.py [hunk 2]
```diff
@@ -1102,3 +1106,13 @@ def isMarshalledIAccessible(IAccessibleObject):
...
+# As winEvents always contain the true thread ID.
+consoleWindowsToThreadIDs = {}
```
Comment (leonardder) [inline source/IAccessibleHandler.py]: Typing

```suggestion
consoleWindowsToThreadIDs: Dict[int, int] = {}
```

File: source/NVDAObjects/UIA/winConsoleUIA.py [hunk 3]
```diff
@@ -254,6 +254,17 @@ class WinConsoleUIA(KeyboardHandlerBasedTypedCharSupport):
 	#: the caret in consoles can take a while to move on Windows 10 1903 and later.
 	_caretMovementTimeoutMultiplier = 1.5
 
+	def _get_windowThreadID(self):
```
Comment (leonardder) [inline source/NVDAObjects/UIA/winConsoleUIA.py]: Not that I think this should happen now, but I wonder whether we want a base class for console windows that the legacy and UIA impls inherit from.

**Review by LeonarddeR — APPROVED** [review]
Only some style comments

**Pull request merged**